client/object_put.go

```diff
@@ -325,3 +334,173 @@ func (c *Client) ObjectPutInit(ctx context.Context, hdr object.Object, signer us
+//
+// Return errors:
+//   - [apistatus.ErrServerInternal]: internal server error described in the text message;
+//   - [apistatus.ErrObjectAccessDenied]: the signer does not authenticate any
+//     NeoFS storage node matching storage policy of the container referenced by the
+//     replicated object;
+//   - [apistatus.ErrContainerNotFound]: the container to which the replicated
+//     object is associated was not found.
+func (c *Client) ReplicateObject(ctx context.Context, src io.ReadSeeker, signer neofscrypto.Signer) error {
+	const svcName = "neo.fs.v2.object.ObjectService"
+	const opName = "Replicate"
+	stream, err := c.c.Init(common.CallMethodInfoUnary(svcName, opName),
+		client.WithContext(ctx), client.AllowBinarySendingOnly())
+	if err != nil {
+		return fmt.Errorf("init service=%s/op=%s RPC: %w", svcName, opName, err)
+	}
+
+	msg, err := prepareReplicateMessage(src, signer)
+	if err != nil {
+		return err
+	}
+
+	err = stream.WriteMessage(client.BinaryMessage(msg))
+	if err != nil && !errors.Is(err, io.EOF) { // io.EOF means the server closed the stream on its side
+		return fmt.Errorf("send request: %w", err)
+	}
+
+	var resp replicateResponse
+	err = stream.ReadMessage(&resp)
+	if err != nil {
+		if errors.Is(err, io.EOF) {
+			err = io.ErrUnexpectedEOF
+		}
+
+		return fmt.Errorf("recv response: %w", err)
+	}
+
+	_ = stream.Close()
+
+	return resp.err
+}
+
+func prepareReplicateMessage(src io.ReadSeeker, signer neofscrypto.Signer) ([]byte, error) {
+	var objSize uint64
+	switch v := src.(type) {
+	default:
+		n, err := src.Seek(0, io.SeekEnd)
+		if err != nil {
+			return nil, fmt.Errorf("seek to end: %w", err)
+		} else if n < 0 {
+			return nil, fmt.Errorf("seek to end returned negative value %d", objSize)
+		}
+
+		_, err = src.Seek(-n, io.SeekCurrent)
+		if err != nil {
+			return nil, fmt.Errorf("seek back to initial pos: %w", err)
+		}
+
+		objSize = uint64(n)
+	case *os.File:
+		fileInfo, err := v.Stat()
+		if err != nil {
+			return nil, fmt.Errorf("get file info: %w", err)
+		}
+
+		objSize = uint64(fileInfo.Size())
+	case *bytes.Reader:
+		n := v.Size()
+		if objSize < 0 {
+			return nil, fmt.Errorf("negative byte buffer size return %d", objSize)
+		}
+
+		objSize = uint64(n)
+	}
+
+	// TODO: limit the objSize?
+
+	// calculate template signature to know its size
+	sigTmpl, err := signer.Sign(nil)
+	if err != nil {
+		return nil, fmt.Errorf("calculate signature of empty ata: %w", err)
+	}
+
+	bPubKey := neofscrypto.PublicKeyBytes(signer.Public())
+	sigScheme := uint64(signer.Scheme())
+
+	const fieldNumObject = 1
+	const fieldNumSignature = 2
+
+	sigSize := protowire.SizeTag(fieldNumSigPubKey) + protowire.SizeBytes(len(bPubKey)) +
+		protowire.SizeTag(fieldNumSigVal) + +protowire.SizeBytes(len(sigTmpl)) +
+		protowire.SizeTag(fieldNumSigScheme) + protowire.SizeVarint(sigScheme)
+
+	msgSize := protowire.SizeTag(fieldNumObject) + protowire.SizeVarint(objSize) +
+		protowire.SizeTag(fieldNumSignature) + sigSize
```

Comment:
> this code just kills many of protobuf features

yeah, we _can_ leverage protobuf power with transparent mechs and all this can be improved, we just don’t have enough codebase for this. Anyway, current approach is more efficient then previously used decode+Put